src/engine.rs

```diff
@@ -39,10 +45,13 @@ pub(crate) fn reformat(
     }
 
     let spacing_diff = model.into_diff();
-    if let Some(explanation) = &mut explanation {
+    if let ExtraInfo::Explanation(explanation) = &mut extra_info {
         if spacing_diff.has_changes() {
             explanation.extend(spacing_diff.edits.clone())
         }
+    } else if let ExtraInfo::Edits(spacing_edits, _) = &mut extra_info {
+        spacing_edits
+            .extend(spacing_diff.edits.iter().map(|(ae, _)| ae.clone()).collect::<Vec<_>>());
```

Comment:
I'm not sure if I understand why we can't re-use the `Explanation`-structure here (i.e. why are two vecs even needed)?

Reply:
The two vecs have to be separate because the delete ranges of the first vec of edits refer to indices in the original text, while those in the second vec refer apply to the text after the first vec of edits have been applied. That's why the unreasonnably complicated `textedit_merge` algorithm is needed in [nix-community/rnix-lsp#87](https://github.com/nix-community/rnix-lsp/pull/87). If they're not returned separately, there's no way of being certain where the first ends and the second begins.

Reply:
Pretty much, though its actually a problem when there are any edits in the first set that affect indices before any of the edits in the second set, because (from my understanding) edits' delete ranges are based on byte indices in the text string, not line and column coordinates.

Also, sorry that I haven't gotten to updating the tests on the other PR yet, I'll try to get that done soon.

Reply:
> delete ranges are based on byte indices in the text string, not line and column coordinates.

I'd recommend to look the details up in the spec for the lsp-related stuff. We had issues like this in the past: https://github.com/nix-community/rnix-lsp/pull/50

Reply:
That's a good point, I did a quick test with a file containing multi-byte characters, and it seems to behave correctly, but we should probably add a test for this kind of thing at some point.